firmware/baseband/dsp_decimate.cpp

```diff
@@ -203,7 +203,73 @@ buffer_c16_t FIRC8xR16x24FS4Decim4::execute(
 
     const auto k = output_scale;
 
-    const size_t count = src.count / decimation_factor;
+    size_t count = src.count / decimation_factor;   // 2048 buffer_c8 input src samples / 4 = 512 buffer c16 output samples.
+    /*
+    if (count > 256) {                  // Just investigation test .
+            count =256;
+    }
+    */
+
+    for (size_t i = 0; i < count; i++) {
+        const vec4_s8* const in = static_cast<const vec4_s8*>(__builtin_assume_aligned(&src.p[i * decimation_factor], 4));
+
+        complex32_t accum;
+
+        // Oldest samples are discarded.
+        accum = mac_fs4_shift(z, t, 0, accum);
+        accum = mac_fs4_shift(z, t, 1, accum);
+
+        // Middle samples are shifted earlier in the "z" delay buffer.
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 0, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 1, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 2, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 3, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 4, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 5, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 6, accum);
+        accum = mac_fs4_shift_and_store(z, t, decimation_factor, 7, accum);
+
+        // Newest samples come from "in" buffer, are copied to "z" delay buffer.
+        accum = mac_fs4_shift_and_store_new_c8_samples(z, t, in, decimation_factor, 0, taps_count, accum);
+        accum = mac_fs4_shift_and_store_new_c8_samples(z, t, in, decimation_factor, 1, taps_count, accum);
+
+        d[i] = scale_round_and_pack(accum, k);
+    }
+
+    return {
+        dst.p,
+        count,
+        src.sampling_rate / decimation_factor};
+}
+
+
+
+// FIRC8xR16x24FS4Decim4_256 //////////////////////////////////////////////////
+// Cloned from previous FIRC8xR16x24FS4Decim4 limiting the output buffer C16 samples to 256
+// Waterfall fft function was designed to handle as a max. 256 Complex samples , (2048 /8 = 256), 
+// if we just use Decimation /4 , we got 2048 c8 src input  / 4 = 512 C16 Complex output  samples and Waterfall crashes.
+
+void FIRC8xR16x24FS4Decim4_256::configure(
+    const std::array<tap_t, taps_count>& taps,
+    const int32_t scale,
+    const Shift shift) {
+    taps_copy(taps.data(), taps_.data(), taps_.size(), shift == Shift::Up);
+    output_scale = scale;
+    z_.fill({});
+}
+
+buffer_c16_t FIRC8xR16x24FS4Decim4_256::execute(
```

Comment:
I haven't looked too closely, but I don't think this is what we need. It would probably be better to just copy every other value from the 512 buffer into the 256 buffer in proc_capture before updating the waterfall. I don't think we need a new type.

Reply:
Ok,  we can discuss,  but at the moment I still do not have a clear image about how the waterfall collects the output buffer dst  of  second decim. block : decim1 to  calculate the FFT , and  eventually to write it to the SD card if we pressed the REC .
I wonder if both processes  feed it from the same output decim0 dst{} , and if  both processes are expecting  a maximum buffer x 256  c16  ,  or if that limitation is only in the waterfall,   ????   I imagine , that as SD card is having a random write latency , and maybe with asyncrhonus write speed ,  we should have  another buffer just for the write SD file process .  (but I still do not have those details ) .    I would appreciate if we can make it clear .